### ISSUE TITLE
Complications with Chunk by Chunk mod - a "don't teleport just point to dimension"-blacklist might help?

### ISSUE THREAD

**Issue by xFirefalconx** [Corail31/tombstone_lite#197]
### Minecraft Version

1.18.2 (edited)

### Forge Version

40.2.17

### Corail Tombstone Version

7.6.6

### What happened?

I could not select 1.18.2 this is for that minecraft version - but it might be on other MC versions, too (including 1.20.x, 1.19.x and 1.16.x). If you can confirm this and port a solution to 1.19.4 for example, I'd offer my help to port it to 1.18.2 and make a PR if you allow me to. I can cherrypick things so you save time. I'm just bad at implementing full solutions myself, since I also don't know your mod in and out. 

About the issue, if you play Corail Tombstone together with Chunk by Chunk (tested on 1.18.2 with CBC 2.1.1) and create a petrified world, then find a lost tablet. That... sounds cool, but is not ideal. There are a few with the NBT data "minecraft:overworld_snow_gen" (should be redirected to "minecraft:overworld") and "chunkbychunk:netherchunkgeneration" (should be not teleporting you at all, redirecting doesn't work because you fall into a lava pit and burn to a crisp). 

If having those NBT, it teleports you out of the dimension you are supposed to play in, because those dimensions are invalid to spawn in. CBC dev said this is possibly an issue with Corail, that's why I am here. 

It's fine to have them generated in general, good way pointers, but then the tablets should have a blacklist to not be actually able to teleport you, just be a pointer towards a building. 
I can try to kubejs my way through somehow (check the inventory each time a change happens, replace the charged tablet with the uncharged one and warn the player with a chat message) but I am bad at it, it probably makes things more hairy, is error prone, might negatively impact performance and I feel like this might also apply to other mod packs that change dimensions or try to restrict you in that way. 
I tried to cancel/redirect the teleport with scripts already, which failed. The scripts don't even recognize the teleport, somehow. 

If I could, I would have a config list of restricted and redirected dimensions. 
Something along the lines of (pseudocode)
```js
if (tablet.NBT.Contains(RedirectConfigList[])){
tablet.NBT.ReplaceWith(RedirectTarget[]);
}
else if (tablet.NBT.Contains(CancelConfigList)){
tablet.ChargeEvent.Cancel();
chat.SendMessage("Cannot charge with a soul! This tablet seems to be irreversably broken. You can go walk to the marked destination by foot, though!")
}

```

I hope you can look into this. As I said, I'd happily give porting it a try, if you still have the 1.18.2 branch... 

### Gametype

Singleplayer, Dedicated Server

### Happen by testing with only Corail Tombstone (required)

No

### Occurence

occurs all the time

### Other relevant mods

Chunk by Chunk 2.1.1 (2.2.0 should yield same result)

### Log Link

https://gist.github.com/xFirefalconx/9945ce37cededf04a392394f344eb016

### Tested without any Forge "alternatives" (Magma, Mohist, and so on)

- [X] Yes

### Valid Launcher

- [X] Yes

**Comment by Corail31** [comment]
I will test the mod with mine as i'm not sure to understand your issue. Can you give me a step by step to make this happening ?
If the problem is to find a better Y pos for some structures, i can look at it.
If the problem is to prevent some structures to be used, i could reset the tablet with a patch.

**Comment by xFirefalconx** [comment]
Alright! Here's a step by step guide:

1. Install both Chunk By Chunk and Corail Tombstone
2. Run the game
3. Go to the Chunk by Chunk configs and check the setup as following:
- the world sealing is true
- the world gets petrified/turned into bedrock
4. if that is all set, create a new world, it's essential for this step
5. you spawn in minecraft:overworld but it's petrified! the reference dimension(s) are on other dimension layers. please ask further if you need more info on that specifically. 
6. grab a large amount of lost tablets and wait for them to find a location with NBT/dimension data that doesn't fit "minecraft:overworld" *exactly* or "minecraft:nether". If they are for example "minecraft:overworld_snow_gen" then when you teleport, it will change dimension! If you don't change dimension with a tablet, toss it, it's not needed for the reproduction of the issue. 
7. selected tablets may be charged/activated in any way you prefer, it doesn't matter, use a grave stone that charged naturally or with the item to charge it.
8. hold right click and wait for "LOADING" to appear while you don't go to the sealed nether or sealed overworld dimension - you likely found a tablet that breaks my modpack. 
9. if the world around you looks normal, no bedrocked chunks around you or you are not floating in pure lava (in the nether) you found a way to bypass chunk by chunk rules and can now farm anything you want as an user, making the restrictions of the spawn dimensions void (which is the main problem here). it is also unknown to me how you can get back easily without a portal or a charged tablet, again. 
 
Some more info/explanation/pov:
The tablets can easily point to the location as the sealed ones are copies of the actual firstly generated overworld, that is fine. But they should not teleport you to another dimension if it's just the reference layer for chunk by chunk to grab chunk data you can recover, they should check for the same structure reference in the minecraft:overworld or minecraft:nether (I am not sure about the typing of the nether here, just about overworld) in this case. 
 
The purpose of chunk by chunk is to rebuild a petrified/lost world by copying the chunks from the original generation to the petrified overworld.

**Comment by Corail31** [comment]
If you're starting with only one chunk, wouldn't the real solution be to disable Lost Tablets (as any location found by those Lost Tablets would be a chunk that you recreated and visited) ? or i missed a detail of the other mod.

**Comment by xFirefalconx** [comment]
"(as any location found by those Lost Tablets would be a chunk that you recreated and visited)" I am not sure I understand your phrasing here (if you can, please elaborate how you mean that?), I can sum it up this way though if that may be some help: 

The general playthrough is that you start at that location (in my pack's case 4 chunks), yes. But it's up to the player whether they stay there or go explore the world and spawn new chunks elswhere. I also recall the dev or some player suggesting players spawning in spreaded places so everybody got their own battle to fight.
You don't have to unlock any chunk close to your start, it's just your spawn. Some people may find Supplemantaries' sign towards a new village and expand and settle there, others may be searching for a specific biome with Nature's Compass (and teleport there), which also works flawlessly. Maybe some want to find treasure troves, so they could use lost tablets to get there. They first have to discover their existence, too, so they might not use that method until later, and even if they do from the start, that's fine. It's about the aspect of limited resources and unlocking new content, by your own effort. 
If you have been unlocking chunks for a long time and then suddenly find a really easy way to teleport to an unpetrified world with an abundance of resources without too much effort, why would you stay in the petrified world? It also destroys the chunk by chunk unpetrification, because those chunks touched on the backup dimension might be altered, thus things missing or things added to the chunk by chunk world you wouldn't want to appear there when you want to restore those chunks later. 
Knowledge of those places is fine. People might be encouraged to venture further if they know there are tools to find hidden places. Reduces frustration, makes it more exciting to go look for new places. 

I also have the option in my pack to start in a normal world and go a different route during the playthrough. Having full access to lost tablets there would be a good thing, actually. (I can alter configs based on button setups so having different configs for each world is fine here)

**Comment by Corail31** [comment]
@xFirefalconx I didn't find the config "the world gets petrified/turned into bedrock".
I tried with ChunkByChunk-1.18.2-2.2.0 for Forge.
https://github.com/immortius/chunkbychunk/issues/89

**Comment by xFirefalconx** [comment]
My apologies, I misremembered the name of the setting. I looked into the config file (in defaultconfigs), it says as following:

```
[ChunkGeneration]
	#Is ChunkByChunk generation enabled
	enabled = true
	#Should empty chunks be generated as bedrock
	seal_world = true
``` 
- this is how I set it up in my specific case and how it will be visible the best

**Comment by Corail31** [comment]
For a provisory solution, you can change the configs : "lost_tablet_search_outside_world", "lost_tablet_modded_structure" and "lostTabletDeniedStructures".
For the final solution, i guess i can avoid any dimension with a namespace "chunkbychunk" or with "overworld:overworld_[something]_gen".
@xFirefalconx contact me on CurseForge to provide you a version of test ;)

**Comment by xFirefalconx** [comment]
Thank you! For the provisory solution, I checked the configs but wasn't able to find such categories on 7.6.6 - either I look the wrong page or they don't exist there? 
I am very, very grateful you take the time to improve on this situation. 
Regarding the test version, I sent you a message on Curseforge a few minutes ago :)

**Comment by Corail31** [comment]
You need to look the configs in the config folder. My config in game is only for client configs and player preferences.
I'll publish the new version as soon as you confirm me it works properly :)

**Comment by xFirefalconx** [comment]
Observation: I had a tablet with "minecraft:the_nether" just now which I charged and then it said there's no way to appear there - that's fine to me in this case but maybe worth looking into as to why that happens. Is the second one I discovered doing that. 
Other than that, I had two hotbars full with tablets, none if which showed the false dimensions yet. I give a cautious "seems fine" from my side.
